src/include/pka_cpu.h

```diff
@@ -79,17 +85,212 @@ static inline uint64_t pka_early_cpu_speed(void)
+        PKA_ERROR(PKA_CPU, "\nOpening file failed with error:%d.\n", errno);
+        return NULL;
+    }
+
+    if (!fstat(fd, &sbuf)) {
+        *file_len = sbuf.st_size;
+    }
+
+    file_mem = malloc(*file_len);
+    if (file_mem == NULL) {
+        PKA_ERROR(PKA_CPU, "\nMemory alloc failed when reading file.\n");
+        goto out;
+    }
+
+    while (read_count != *file_len && read_status != 0) {
+        read_status = read(fd, file_mem + read_count, *file_len - read_count);
+        if (read_status == -1) {
+            PKA_ERROR(PKA_CPU, "\nReading file failed.\n");
+            goto err;
+        }
+        else
+            read_count += read_status;
+    }
+
+    if (read_count != *file_len) {
+        PKA_ERROR(PKA_CPU, "\n EOF reached before expected length.\n");
+        goto err;
+    }
+    else
+        goto out;
+
+err:
+    free(file_mem);
+    file_mem = NULL;
+out:
+    if (close(fd) == -EPERM)
+        PKA_ERROR(PKA_CPU, "\nClosing the file failed with error=%d.\n", errno);
+
+    return file_mem;
+}
+
+static int pka_cpu_checksum(const uint8_t *buf, size_t len)
+{
+    uint8_t sum;
+    size_t  d;
+
+    sum = 0;
+
+    for (d = 0; d < len; d++)
+         sum += buf[d];
+
+    return sum == 0;
+}
+
+static uint16_t pka_cpu_read_dmi_table(const char *sysfs)
+{
+    uint8_t          *buf, *data, *next;
+    pka_cpu_dmi_hdr_t dmi_hdr;
+    uint32_t          file_len;
+    size_t            file_size;
+
+    buf = pka_cpu_read_sysfs(&file_size, sysfs);
+    if (buf == NULL)
+        return 0;
+
+    file_len = file_size;
+    data     = buf;
+
+    while (data + PKA_DMI_TABLE_LEN <= buf + file_len) {
+
+        /// Copy the header information
+        dmi_hdr.type   = data[PKA_DMI_HEADER_TYPE_OFFSET];
+        dmi_hdr.length = data[PKA_DMI_HEADER_LEN_OFFSET];
+        dmi_hdr.handle = WORD(data + PKA_DMI_HEADER_HANDLE_OFFSET);
+        dmi_hdr.data   = data;
+
+        if (dmi_hdr.length < PKA_DMI_TABLE_LEN) {
+            PKA_ERROR(PKA_CPU, "\nInvalid entry length, DMI table is broken.\n");
+            break;
+        }
+
+        if (dmi_hdr.type == PKA_DMI_TYPE_END)
+            break;
+
+        next = data + dmi_hdr.length;
+        while ((unsigned long)(next - buf + 1) < file_len
+            && (next[0] != 0 || next[1] != 0))
+                next++;
+
+        next += 2;
+
+        if ((unsigned long)(next - buf) > file_len) {
+            PKA_ERROR(PKA_CPU, "\nDMI structure is truncated.\n");
+            break;
+        }
+
+        /// Processor information
+        if (dmi_hdr.type == PKA_DMI_PROC_TYPE) {
+            if (dmi_hdr.length < PKA_DMI_PROC_TABLE_LEN) {
+                PKA_ERROR(PKA_CPU, "\nIncomplete processor information.\n");
+                return 0;
+            }
+            /// Read frequency from offset
+            return WORD(dmi_hdr.data + PKA_DMI_PROC_FREQ_OFFSET);
+        }
+
+        data = next;
+    }
+    return 0;
+}
+
+static uint64_t pka_cpu_get_freq_from_smbios()

```

Comment:
It is better that you declare it as:

static uint64_t pka_cpu_get_freq_from_smbios(void)